_posts/2017-12-21-totw-42.md

```diff
@@ -0,0 +1,92 @@
+Calvin Coolidge*
+
+One of the main difficulties with writing constructors in C++ (at Google) is
+that they effectively must succeed, because they have no way to report failure
+to the caller. (Google's C++ style forbids exceptions.) You can use an
+`abort()`, of course, but doing so crashes the whole program, which is often
+ unacceptable in production code.
+
+If your class’s initialization logic can’t avoid the possibility of failure, one
+common approach is to give the class an initializer method (also called an "init
+method"), which performs any initialization work that might fail, and signals
+that failure via its return value. The assumption is usually that the user will
+call this method immediately after construction, and if it fails the user will
+immediately destroy the object. However, these assumptions are not always
+documented, nor always obeyed. It’s all too easy for users to start calling
+other methods before initialization, or after initialization has failed.
+Sometimes the class actually encourages this behavior, e.g. by providing methods
+to configure the object before initializing it, or to read errors out of it
+after initialization fails.
+
+This design commits you to maintaining a class with at least two distinct
+user-visible states, and often three: initialized, uninitialized, and
+initialization-failed. Making such a design work requires a lot of discipline:
+every method of the class has to specify what states it can be called in, and
+users have to comply with these rules. If this discipline lapses, client
+developers will tend to write whatever code happens to work, regardless of
+what you intended to support. When that starts to happen, maintainability
+nosedives, because your implementation has to support whatever combinations of
+pre-initialization method calls your clients have started to depend on. In
+effect, your implementation has become your interface.
+
+Fortunately, there’s a simple alternative that lacks these drawbacks: provide a
+*factory function* which creates and initializes instances of your class, and
+returns them by pointer, using null to indicate failure. Here’s a toy example:
+
+```c++
+// foo.h
+class Foo {
+ public:
+  // Factory method: creates and returns a Foo.
+  // May return null on failure.
+  static std::unique_ptr<Foo> Create();
+
+  // Foo is not copyable.
+  Foo(const Foo&) = delete;
+  Foo& operator=(const Foo&) = delete;
+
+ private:
+  // Clients can't invoke the constructor directly.
+  Foo()
+};
+
+// foo.c
+std::unique_ptr<Foo> Foo::Create() {
+  // Note that since Foo's constructor is private, we have to use new.
+  return absl::WrapUnique(new Foo());
+}
+```
+
+In many cases, this pattern gives you the best of both worlds: the factory
+function `Foo::Create()` exposes only fully-initialized objects like a
+constructor, but it can indicate failure like an initializer method. Another
+advantage of factory functions is that they can return instances of any subclass
```

Comment:
in the unique_ptr case (not for optional)

Reply:
Done.

However, I feel like we now need an absl::optional example (and the tip may not warrant that level of detail). Maybe there's a one-liner you can fit in the current example as an alternative?

Also, please look over the last paragraph re: absl::optional. It's already a "value type" return type, no?

We may need to re-write this as two sections: factory functions using unique_ptr and factory functions using optionals.

Reply:
I think it's good enough as is.